### PR TITLE
Add scripts to perform automated benchmarking against Open Force Field protein-ligand benchmarks

### LINKED ISSUE #854: Benchmarking and evaluation scripts
We should pull in [automated scripts for running the free energy benchmark](https://www.choderalab.org/publications/2021/7/5/best-practices-for-constructing-preparing-and-evaluating-protein-ligand-binding-affinity-benchmarks) into perses.

### PR DESCRIPTION
**Pull request by ijpulidos**
## Description

Introduces a command line utility to automatically run benchmarks using the data set found in https://github.com/openforcefield/protein-ligand-benchmark/ 

## Motivation and context

The idea is having a way to automatically monitor and validate real-life free energy calculations results when we change something in perses.

Resolves #854 

## How has this been tested?

Manually tested (it is intended to be run manually but we can think about writing some automated tests for it if needed)

## Change log

```
CLI tool for running standard protein-ligand benchmarks from https://github.com/openforcefield/protein-ligand-benchmark/
```

# TODO

- [x] automate executing benchmarks from https://github.com/openforcefield/protein-ligand-benchmark
- [x] robust way for requesting files (http request)
- ~[ ] Specify forcefields xml files to use via cli~ (Pushed for a future API enhancement release with this change)
- [x] Basic automated analysis of results
- [x] Analysis plots/visualizations

### CODE REVIEW

File: benchmarks/benchmark_analysis.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+
+# Generate list with simulation objects
+simulations = get_simulations_data(out_dirs)
+
+# Generate csv file
+csv_path = f'./{target}_arsenic.csv'
+to_arsenic_csv(ligands_dict, simulations, out_csv=csv_path)
+
+
+# TODO: Separate plotting in a different file
```
Comment (jchodera) [inline benchmarks/benchmark_analysis.py:195]: @ijpulidos : I think we still want to separate the plotting into a different script!

**Comment by jchodera** [comment]
A few comments after trying to run this!

Would be great to:
* Have this organize the files it generates into directories
* Neatly organize the distilled CSV data for plotting into a separate directory and separate plotting script
* Provide a `cleanup.sh` script that cleans up all the files generated

**Comment by jchodera** [comment]
@mikemhenry : Looks like we need to add `arsenic` to the dependencies upon release:
```
$ python benchmark_analysis.py 
Warning: importing 'simtk.openmm' is deprecated.  Import 'openmm' instead.
hostname
Traceback (most recent call last):
  File "/lila/data/chodera/chodera/perses/perses/benchmarks/benchmark_analysis.py", line 22, in <module>
    from openff.arsenic import plotting, wrangle
ModuleNotFoundError: No module named 'openff.arsenic'
```
I'll add this to the dev environment, but let's be sure to add it to the conda-forge recipe feedstock.

**Comment by mikemhenry** [comment]
Yes, I'll also add it to 0.9.2, then if we ever need to use that version to check a benchmark it will be easier to install

**Comment by jchodera** [comment]
Some of these jobs appear to have crashed when they made it to the `complex` phase with this puzzling error:
```
022-01-29 01:39:31,361:(0.00s):openmmtools.multistate.multistatesampler:********************************************************************************
2022-01-29 01:39:31,362:(0.00s):openmmtools.multistate.multistatesampler:Iteration 1/5000
2022-01-29 01:39:31,362:(0.00s):openmmtools.multistate.multistatesampler:********************************************************************************
2022-01-29 01:39:31,362:(0.00s):mpiplus.mpiplus:Single node: executing <function ReplicaExchangeSampler._mix_replicas at 0x2ac7fe2043a0>
2022-01-29 01:39:31,362:(0.00s):openmmtools.multistate.replicaexchange:Mixing replicas...
2022-01-29 01:39:31,362:(0.00s):openmmtools.utils:Mixing of replicas took    0.001s
2022-01-29 01:39:31,363:(0.00s):openmmtools.multistate.replicaexchange:Accepted 8568/16000 attempted swaps (53.5%)
2022-01-29 01:39:31,363:(0.00s):openmmtools.multistate.multistatesampler:Propagating all replicas...
2022-01-29 01:39:31,363:(0.00s):mpiplus.mpiplus:Running _propagate_replica serially.
Traceback (most recent call last):
  File "/lila/home/chodera/miniconda/envs/perses-dev/lib/python3.9/site-packages/openmmtools/cache.py", line 445, in get_context
    context = self._lru[context_id]
  File "/lila/home/chodera/miniconda/envs/perses-dev/lib/python3.9/site-packages/openmmtools/cache.py", line 147, in __getitem__
    entry = self._data.pop(key)
KeyError: (-2109681550591862395, 1542409512075896928)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/lila/data/chodera/chodera/perses/perses/benchmarks/run_benchmarks.py", line 176, in <module>
    run_relative_perturbation(lig_a_index, lig_b_index, reverse=is_reversed)
  File "/lila/data/chodera/chodera/perses/perses/benchmarks/run_benchmarks.py", line 92, in run_relative_perturbation
    run(new_yaml)
  File "/lila/home/chodera/miniconda/envs/perses-dev/lib/python3.9/site-packages/perses/app/setup_relative_calculation.py", line 911, in run
    hss_run.extend(setup_options['n_cycles'])
  File "/lila/home/chodera/miniconda/envs/perses-dev/lib/python3.9/site-packages/openmmtools/multistate/multistatesampler.py", line 728, in extend
    self.run(n_iterations)
  File "/lila/home/chodera/miniconda/envs/perses-dev/lib/python3.9/site-packages/openmmtools/multistate/multistatesampler.py", line 681, in run
    self._propagate_replicas()
  File "/lila/home/chodera/miniconda/envs/perses-dev/lib/python3.9/site-packages/openmmtools/utils.py", line 90, in _wrapper
    return func(*args, **kwargs)
  File "/lila/home/chodera/miniconda/envs/perses-dev/lib/python3.9/site-packages/openmmtools/multistate/multistatesampler.py", line 1196, in _propagate_replicas
    propagated_states, replica_ids = mpiplus.distribute(self._propagate_replica, range(self.n_replicas),
  File "/lila/home/chodera/miniconda/envs/perses-dev/lib/python3.9/site-packages/mpiplus/mpiplus.py", line 512, in distribute
    all_results = [task(job_args, *other_args, **kwargs) for job_args in distributed_args]
  File "/lila/home/chodera/miniconda/envs/perses-dev/lib/python3.9/site-packages/mpiplus/mpiplus.py", line 512, in <listcomp>
    all_results = [task(job_args, *other_args, **kwargs) for job_args in distributed_args]
  File "/lila/home/chodera/miniconda/envs/perses-dev/lib/python3.9/site-packages/openmmtools/multistate/multistatesampler.py", line 1223, in _propagate_replica
    mcmc_move.apply(thermodynamic_state, sampler_state)
  File "/lila/home/chodera/miniconda/envs/perses-dev/lib/python3.9/site-packages/openmmtools/mcmc.py", line 1114, in apply
    super(LangevinDynamicsMove, self).apply(thermodynamic_state, sampler_state)
  File "/lila/home/chodera/miniconda/envs/perses-dev/lib/python3.9/site-packages/openmmtools/mcmc.py", line 655, in apply
    context, integrator = context_cache.get_context(thermodynamic_state, integrator)
  File "/lila/home/chodera/miniconda/envs/perses-dev/lib/python3.9/site-packages/openmmtools/cache.py", line 447, in get_context
    context = thermodynamic_state.create_context(integrator, self._platform, self._platform_properties)
  File "/lila/home/chodera/miniconda/envs/perses-dev/lib/python3.9/site-packages/openmmtools/states.py", line 1172, in create_context
    return openmm.Context(system, integrator)
  File "/lila/home/chodera/miniconda/envs/perses-dev/lib/python3.9/site-packages/openmm/openmm.py", line 16305, in __init__
    _openmm.Context_swiginit(self, _openmm.new_Context(*args))
openmm.OpenMMException: Error initializing FFT: 5
```
I'll investigate further!

**Comment by jchodera** [comment]
I've run into another stochastic failure:
```
2022-01-29 20:31:00,950:(0.00s):openmmtools.multistate.multistatesampler:Iteration 2036/5000
2022-01-29 20:31:00,950:(0.00s):openmmtools.multistate.multistatesampler:********************************************************************************
2022-01-29 20:31:00,951:(0.00s):mpiplus.mpiplus:Single node: executing <function ReplicaExchangeSampler._mix_replicas at 0x2b435e754430>
2022-01-29 20:31:00,951:(0.00s):openmmtools.multistate.replicaexchange:Mixing replicas...
2022-01-29 20:31:00,951:(0.00s):openmmtools.utils:Mixing of replicas took    0.000s
2022-01-29 20:31:00,951:(0.00s):openmmtools.multistate.replicaexchange:Accepted 458/3456 attempted swaps (13.3%)
2022-01-29 20:31:00,951:(0.00s):openmmtools.multistate.multistatesampler:Propagating all replicas...
2022-01-29 20:31:00,951:(0.00s):mpiplus.mpiplus:Running _propagate_replica serially.
2022-01-29 20:31:01,097:(0.15s):openmmtools.mcmc:Potential energy is NaN after 0 attempts of integration with move LangevinSplittingDynamicsMove Attempting a restart...
Traceback (most recent call last):
  File "/lila/data/chodera/chodera/perses/perses/benchmarks/run_benchmarks.py", line 176, in <module>
    run_relative_perturbation(lig_a_index, lig_b_index, reverse=is_reversed)
  File "/lila/data/chodera/chodera/perses/perses/benchmarks/run_benchmarks.py", line 92, in run_relative_perturbation
    run(new_yaml)
  File "/lila/home/chodera/miniconda/envs/perses-dev/lib/python3.9/site-packages/perses/app/setup_relative_calculation.py", line 911, in run
    hss_run.extend(setup_options['n_cycles'])
  File "/lila/home/chodera/miniconda/envs/perses-dev/lib/python3.9/site-packages/openmmtools/multistate/multistatesampler.py", line 728, in extend
    self.run(n_iterations)
  File "/lila/home/chodera/miniconda/envs/perses-dev/lib/python3.9/site-packages/openmmtools/multistate/multistatesampler.py", line 681, in run
    self._propagate_replicas()
  File "/lila/home/chodera/miniconda/envs/perses-dev/lib/python3.9/site-packages/openmmtools/utils.py", line 90, in _wrapper
    return func(*args, **kwargs)
  File "/lila/home/chodera/miniconda/envs/perses-dev/lib/python3.9/site-packages/openmmtools/multistate/multistatesampler.py", line 1196, in _propagate_replicas
    propagated_states, replica_ids = mpiplus.distribute(self._propagate_replica, range(self.n_replicas),
  File "/lila/home/chodera/miniconda/envs/perses-dev/lib/python3.9/site-packages/mpiplus/mpiplus.py", line 512, in distribute
    all_results = [task(job_args, *other_args, **kwargs) for job_args in distributed_args]
  File "/lila/home/chodera/miniconda/envs/perses-dev/lib/python3.9/site-packages/mpiplus/mpiplus.py", line 512, in <listcomp>
    all_results = [task(job_args, *other_args, **kwargs) for job_args in distributed_args]
  File "/lila/home/chodera/miniconda/envs/perses-dev/lib/python3.9/site-packages/openmmtools/multistate/multistatesampler.py", line 1223, in _propagate_replica
    mcmc_move.apply(thermodynamic_state, sampler_state)
  File "/lila/home/chodera/miniconda/envs/perses-dev/lib/python3.9/site-packages/openmmtools/mcmc.py", line 1114, in apply
    super(LangevinDynamicsMove, self).apply(thermodynamic_state, sampler_state)
  File "/lila/home/chodera/miniconda/envs/perses-dev/lib/python3.9/site-packages/openmmtools/mcmc.py", line 664, in apply
    sampler_state.apply_to_context(context, ignore_velocities=self.reassign_velocities)
  File "/lila/home/chodera/miniconda/envs/perses-dev/lib/python3.9/site-packages/openmmtools/states.py", line 2272, in apply_to_context
    context.setPositions(self._unitless_positions)
  File "/lila/home/chodera/miniconda/envs/perses-dev/lib/python3.9/site-packages/openmm/openmm.py", line 16035, in setPositions
    return _openmm.Context_setPositions(self, positions)
openmm.OpenMMException: Error downloading array posq: CUDA_ERROR_ILLEGAL_ADDRESS (700)
terminate called after throwing an instance of 'OpenMM::OpenMMException'
  what():  Error deleting array param1: CUDA_ERROR_ILLEGAL_ADDRESS (700)
/scratch/lsftmp/1643482489.4935758.16.shell: line 30: 60319 Aborted                 (core dumped) python run_benchmarks.py --target tyk2 --edge $(( $LSB_JOBINDEX - 1 ))
```
I think we need to protect against `OpenMMException`s with a `try...except` block. Ideally, we could figure out a way to put this within the `MultiStateSampler` machinery, or even within the `ContextCache` machinery.

**Comment by jchodera** [comment]
@ijpulidos : Is there currently a way to restart calculations that die?

**Comment by ijpulidos** [comment]
@jchodera I just re-run the command and it should automatically pick it up from the previous checkpoint.

**Comment by ijpulidos** [comment]
I've also faced a similar issue with core being dumped in some of my runs. I'm not sure what's causing those.

**Comment by jchodera** [comment]
There appear to be some issues with how the `run_benchmarks.py` script is written: If each script is supposed to handle a transformation, what happens when this code runs concurrently for all transformations?
```python
target_dir = targets_dict[target]['dir']
pdb_url = f"{base_repo_url}/raw/master/data/{target_dir}/01_protein/crd/protein.pdb"
pdb_file = retrieve_file_url(pdb_url)

# Fetch cofactors crystalwater pdb file                                                                                                                                                                                                        
# TODO: This part should be done using plbenchmarks API - once there is a conda pkg                                                                                                                                                            
cofactors_url = f"{base_repo_url}/raw/master/data/{target_dir}/01_protein/crd/cofactors_crystalwater.pdb"
cofactors_file = retrieve_file_url(cofactors_url)

# Concatenate protein with cofactors pdbs                                                                                                                                                                                                      
concatenate_files((pdb_file, cofactors_file), 'target.pdb')

# Fetch ligands sdf files and concatenate them in one                                                                                                                                                                                          
# TODO: This part should be done using plbenchmarks API - once there is a conda pkg                                                                                                                                                            
ligands_url = f"{base_repo_url}/raw/master/data/{target_dir}/00_data/ligands.yml"
with fetch_url_contents(ligands_url) as response:
    ligands_dict = yaml.safe_load(response.read())
ligand_files = []
for ligand in ligands_dict.keys():
    ligand_url = f"{base_repo_url}/raw/master/data/{target_dir}/02_ligands/{ligand}/crd/{ligand}.sdf"
    ligand_file = retrieve_file_url(ligand_url)
    ligand_files.append(ligand_file)
# concatenate sdfs                                                                                                                                                                                                                             
concatenate_files(ligand_files, 'ligands.sdf')
```
Presumably, we want to break this into multiple stages
* retrieve files needed for one or more benchmark system(s)
* set up all transformations in parallel
* run or resume all transformations in parallel
* analyze data to generate plots

We'll want to find a more clever way to refactor this in the next release so we can support benchmarking multiple targets at the same time as well.

**Comment by jchodera** [comment]
Thanks!

> I've also faced a similar issue with core being dumped in some of my runs. I'm not sure what's causing those.

We should
1. Attempt to figure out the underlying cause with the HPC folks. In this case, this was `ls15`, CUDA device `0`. 
2. Implement some level of protection against these failures. One option is to wrap the whole simulation in a `try...except` block. That might be the most conservative, safest option, since we don't know how many CUDA contexts need to be invalidated if this occurs.

**Comment by jchodera** [comment]
After correcting a few bugs in the conversion of experimental data, we're now seeing excellent agreement with experiment when using `openff-2.0.0`. Well done, everyone!
plot_absolute - 5 ns - 12 states
plot_relative - 5 ns - 12 states
[tyk2_arsenic - 5 ns - 12 states.csv](https://github.com/choderalab/perses/files/7966729/tyk2_arsenic.-.5.ns.-.12.states.csv)

**Comment by jchodera** [comment]
It looks like 1 ns/replica for 12 states is also doing very well. The RMSE is only slightly degraded to 0.75 [0.55, 1.01] kcal/mol.
We might want to have our analysis scripts plot the performance for various simulation lengths (e.g. every 1 ns/replica).

plot_absolute - 1 ns - 12 states
plot_relative - 1 ns - 12 states
[tyk2_arsenic - 1 ns - 12 states.csv](https://github.com/choderalab/perses/files/7966750/tyk2_arsenic.-.1.ns.-.12.states.csv)

**Review by jchodera — APPROVED** [review]
OK, this is minimally functionally and confirmed to run!
We will iteratively improve this in future.

Feel free to merge when ready, @ijpulidos, and thanks for putting this together!

**Comment by jchodera** [comment]
We’ll need to do more work remediating protein structures https://github.com/openforcefield/protein-ligand-benchmark/issues/20 before we can do more benchmarking for future releases, but I also managed to run mcl1.

Clearly more work to do here, but results are  concordant with the Gapsys and de Groot paper (Gapsys ΔΔG AUE is 1.26+-0.1 kcal/mol; ours is 1.35 [95% CI: 1.09, 1.68] kcal/mol).

mcl1 - plot_absolute - 5 ns - 12 states
mcl1 - plot_relative - 5 ns - 12 states
[mcl1_arsenic - 5 ns - 12 states.csv](https://github.com/choderalab/perses/files/7972453/mcl1_arsenic.-.5.ns.-.12.states.csv)

**Pull request merged**